x11-misc/gpaste: Add version 3.36

zigford opened the pull request.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,82 @@
+VALA_MIN_API_VERSION="0.30"
+VALA_USE_DEPEND="vapigen"
+
+inherit eutils autotools gnome2-utils vala vcs-snapshot
```
Can you write a meson-based ebuild from scratch, since upstream seems to have moved to using meson?

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,82 @@
+LICENSE="GPL-3"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+applet gnome vala"
```
There are more options than this.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,82 @@
+KEYWORDS="~amd64 ~x86"
+IUSE="+applet gnome vala"
+
+CDEPEND="
```
You don't need this with EAPI-7 and BDEPEND.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,82 @@
+
+src_install() {
+	default
+	prune_libtool_files
```
This command is outdated and you don't even inherit the eclass that defines it.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,82 @@
+	prune_libtool_files
+}
+
+pkg_preinst() {
+	gnome2_schemas_savelist
+}
```
Not needed, this is automatically called.

juippis reviewed (commented):
Hi, 
I know you probably copy-pasted the previous ebuild, but it's grown outdated and we'd like any new ebuilds to match our current quality standards. That being said, you'd have a lot to do here if you wanted this to hit the tree. 

Then you need to properly sign-off **all** your commits, 
https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin

And you can use git message footer to link this PR / commit to any open bugs in bugzilla (I know this package has a version bump request) by using `Bug: bug url`. Please see https://www.gentoo.org/glep/glep-0066.html#commit-messages

zigford commented:
Thanks @juippis for the feedback. I'll give a proper crack at resolving the issues you raised.

juippis commented:
If you're going to do it, I'll give few hinters to start so the meson-based review will be smoother:
I see upstream has options to enable installing bash-completion and zsh-completion files, we always install these in Gentoo without USE flags:
https://projects.gentoo.org/qa/policy-guide/installed-files.html#pg0301

Same goes for systemd, I see upstream has an option for it, but you'll have to check whether the option means **linking** to libsystemd.so or **installing** systemd service files. We always install the service files if they are provided.

Here's how to prune static libs correctly nowadays:
https://devmanual.gentoo.org/eclass-reference/ltprune.eclass/index.html
https://projects.gentoo.org/qa/policy-guide/installed-files.html#pg0303

zigford commented:
Updates
* Moved to meson
* Updated use flags (did not break out x-keybindings as it doesn't change deps or build size)
* Fixed missing required use flag for vala and introspection
* Removed unneeded stripping
* Signed off commits

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,63 @@
+		-Dsystemd=true
+		-Dbash-completion=true
+		-Dzsh-completion=true
+		-Dx-keybinder=true
```
I believe this flag will make it depend on *gnome-keybindings* whatever package provides that.

```
C linker for the host machine: x86_64-pc-linux-gnu-gcc ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/x86_64-pc-linux-gnu-pkg-config (1.7.3)
Run-time dependency appstream-glib found: YES 0.7.17
Run-time dependency gdk-3.0 found: YES 3.24.22
Run-time dependency gdk-pixbuf-2.0 found: YES 2.40.0
Run-time dependency gio-2.0 found: YES 2.64.5
Run-time dependency glib-2.0 found: YES 2.64.5
Run-time dependency gobject-2.0 found: YES 2.64.5
Run-time dependency gtk+-3.0 found: YES 3.24.22
Run-time dependency pango found: YES 1.42.4
Run-time dependency dbus-1 found: YES 1.12.20
Found CMake: /usr/bin/cmake (3.18.2)
Run-time dependency gnome-keybindings found: NO (tried pkgconfig and cmake)

meson.build:34:2: ERROR: Dependency "gnome-keybindings" not found, tried pkgconfig and cmake

A full log can be found at /var/tmp/portage/x11-misc/gpaste-3.36.3/work/gpaste-3.36.3-build/meson-logs/meson-log.txt
 * ERROR: x11-misc/gpaste-3.36.3::testworld failed (configure phase):
```
You can straight DEPEND on it, or make it some sort of *keybinds* USE flag maybe.

zigford replied:
Resolved by including x-keybinder only when gnome use flag is is activated.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,63 @@
+IUSE="introspection gnome vala"
+REQUIRED_USE="vala? ( introspection )"
+
+DEPEND="
+	vala? ( $(vala_depend) )
+	virtual/pkgconfig
+"
+RDEPEND="${DEPEND}
+	gnome? (
+		gnome-base/gnome-shell
+	)
+"
+BDEPEND="
+	dev-libs/appstream-glib
+	>=dev-libs/glib-2.48:2
+	>=dev-libs/gobject-introspection-1.48.0
+	sys-apps/dbus
+	>=x11-libs/gdk-pixbuf-2.34:2
+	>=x11-libs/gtk+-3.20:3
+	x11-libs/libX11
+	x11-libs/libXi
+	gnome? (
+		gnome-base/gnome-control-center:2
+		media-libs/clutter
+		x11-libs/pango
```
Your DEPEND and BDEPEND blocks are the other way around, so just move BDEPEND above RDEPEND and call it DEPEND. Then move your current DEPEND where BDEPEND is now, and call it BDEPEND.

```suggestion
DEPEND="
	dev-libs/appstream-glib
	>=dev-libs/glib-2.48:2
	>=dev-libs/gobject-introspection-1.48.0
	sys-apps/dbus
	>=x11-libs/gdk-pixbuf-2.34:2
	>=x11-libs/gtk+-3.20:3
	x11-libs/libX11
	x11-libs/libXi
	gnome? (
		gnome-base/gnome-control-center:2
		media-libs/clutter
		x11-libs/pango
	)
BDEPEND="
	vala? ( $(vala_depend) )
	virtual/pkgconfig
"
RDEPEND="${DEPEND}
	gnome? (
		gnome-base/gnome-shell
	)
"
```

Rest looks good.

zigford commented:
@juippis many thanks for your suggestions and feedback. I think including keybindings only when the gnome use flag (and therefore gnome-control-center dep) is used will resolve the issue of x-keybinding feature.

Do I need to rebase this against the latest gentoo master?

juippis commented:
> Do I need to rebase this against the latest gentoo master?

Only when there's a merge conflict between your branch and master, I don't believe anyone has touched gpaste in a while.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,64 @@
+		-Dzsh-completion=true
+		$(meson_use introspection introspection)
+		$(meson_use vala vapi)
+		$(meson_use gnome x-keybinder)
```
Regardless, it seems to require gnome-keybinder binary. I have no idea what package provides that, but if you can compile this, you probably know or at least can check it out using *equery* or alike.
```
Calculating dependencies... done!
[ebuild  N     ] x11-misc/gpaste-3.36.3::testworld  USE="-gnome -introspection -vala" 0 KiB

Total: 1 package (1 new), Size of downloads: 0 KiB

Would you like to merge these packages? [Yes/No] y

>>> Verifying ebuild manifests

>>> Emerging (1 of 1) x11-misc/gpaste-3.36.3::testworld
 * gpaste-3.36.3.tar.gz BLAKE2B SHA512 size ;-) ...                                                           [ ok ]
>>> Unpacking source...
>>> Unpacking gpaste-3.36.3.tar.gz to /var/tmp/portage/x11-misc/gpaste-3.36.3/work
>>> Source unpacked in /var/tmp/portage/x11-misc/gpaste-3.36.3/work
>>> Preparing source in /var/tmp/portage/x11-misc/gpaste-3.36.3/work/GPaste-3.36.3 ...
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/x11-misc/gpaste-3.36.3/work/GPaste-3.36.3 ...
meson setup --buildtype plain --libdir lib64 --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --build.pkg-config-path /usr/share/pkgconfig --pkg-config-path /usr/share/pkgconfig --native-file /var/tmp/portage/x11-misc/gpaste-3.36.3/temp/meson.x86_64-pc-linux-gnu.amd64.ini -Dsystemd=true -Dbash-completion=true -Dzsh-completion=true -Dintrospection=false -Dvapi=false -Dx-keybinder=false -Dgnome-shell=false /var/tmp/portage/x11-misc/gpaste-3.36.3/work/GPaste-3.36.3 /var/tmp/portage/x11-misc/gpaste-3.36.3/work/gpaste-3.36.3-build
The Meson build system
Version: 0.55.1
Source dir: /var/tmp/portage/x11-misc/gpaste-3.36.3/work/GPaste-3.36.3
Build dir: /var/tmp/portage/x11-misc/gpaste-3.36.3/work/gpaste-3.36.3-build
Build type: native build
Project name: GPaste
Project version: 3.36.3
C compiler for the host machine: x86_64-pc-linux-gnu-gcc (gcc 10.2.0 "x86_64-pc-linux-gnu-gcc (Gentoo 10.2.0-r1 p2) 10.2.0")
C linker for the host machine: x86_64-pc-linux-gnu-gcc ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/x86_64-pc-linux-gnu-pkg-config (1.7.3)
Run-time dependency appstream-glib found: YES 0.7.17
Run-time dependency gdk-3.0 found: YES 3.24.22
Run-time dependency gdk-pixbuf-2.0 found: YES 2.40.0
Run-time dependency gio-2.0 found: YES 2.64.5
Run-time dependency glib-2.0 found: YES 2.64.5
Run-time dependency gobject-2.0 found: YES 2.64.5
Run-time dependency gtk+-3.0 found: YES 3.24.22
Run-time dependency pango found: YES 1.42.4
Run-time dependency dbus-1 found: YES 1.12.20
Found CMake: /usr/bin/cmake (3.18.2)
Run-time dependency gnome-keybindings found: NO (tried pkgconfig and cmake)

meson.build:34:2: ERROR: Dependency "gnome-keybindings" not found, tried pkgconfig and cmake

A full log can be found at /var/tmp/portage/x11-misc/gpaste-3.36.3/work/gpaste-3.36.3-build/meson-logs/meson-log.txt
 * ERROR: x11-misc/gpaste-3.36.3::testworld failed (configure phase):
 *   (no error message)
```

juippis replied:
https://github.com/Keruspe/GPaste/blob/master/meson.build#L32 and here's the option where the requirement comes from.

zigford replied:
I've worked out that if gnome-control-center is installed it provides the pkg-config file (gnome-keybindings.pc) so that it can find the gnome-keybindings directory (resolves to /usr/share/gnome-control-center/keybindings on my machine).

It does this regardless if gnome-shell option is enabled at compile time or not.

I'm looking for advice, as to which method to resolve:

A) Patch the meson.build file so that it only looks for gnome-keybindings when when the gnome use flag is enabled, and make gnome-control-center a build dep, 

B) Supply the meson option control-center-keybindings-dir with a path (/usr/share/gnome-control-center/keybindings) if gnome use flag is not enabled so that it doesn't attempt to find the gnome-keybindings pkg-config file

thesamesam replied:
So for Option B, that file wouldn't exist (we'd hope), then it'd fail? This wouldn't work if we wanted to disable gnome even if we had the package providing it installed. So we need to do A.

zigford replied:
I'm setting up a VM without all the gnome stuff to test, but at this stage it looks like gnome-control-center is a runtime dependency and is only a build dep in order to determine the keybindings directory. You can suppress it being required as a build dep by specifying the keybindings directory as a configure option.

zigford commented:
I've done some more testing in a VM and have it working quite well with a few tweaks. Hoping this can be signed off and merged, but please let me know if there is anything else that should be fixed.

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,82 @@
+
+src_configure() {
+	local emesonargs=(
+		-Dsystemd=true
```
I am sorry to return this to you... but:
```
Run-time dependency systemd found: NO (tried pkgconfig and cmake)

meson.build:47:4: ERROR: Dependency "systemd" not found, tried pkgconfig and cmake

A full log can be found at /var/tmp/portage/x11-misc/gpaste-3.36.3/work/gpaste-3.36.3-build/meson-logs/meson-log.txt
 * ERROR: x11-misc/gpaste-3.36.3::testworld failed (configure phase):
```
So it does seem to be linking against libsystemd.so. I warned you about it ;)
You'll have to make it an USE flag here. Remember to modify DEPEND+RDEPEND correctly.

Well at least we seem to have moved on from gnome-keybindings!

I warned you it'd be a huge task, but as you might figure out, we don't want to push a broken ebuild to the tree. If we can't compile it, chances are so wouldn't other people.

zigford replied:
All good. My machine and test VM are systemd so did not pick that up. I'll setup a non-systemd machine to verify my changes and get back

juippis commented on x11-misc/gpaste/gpaste-3.36.3.ebuild:
```diff
@@ -0,0 +1,88 @@
+	>=x11-libs/gtk+-3.20:3
+	x11-libs/libX11
+	x11-libs/libXi
+	gnome? (
+		media-libs/clutter
+		x11-libs/pango
+	)
```
This fails with `-gnome`

I'm basically trying to merge this on an empty stage3 image with first `-gnome -vala -introspection -systemd`. Since I don't have any of those enabled on my desktop either.

And when that works, I'd try with every USE flag enabled except systemd. Just FYI how you should test this :) 

```
Run-time dependency mutter-clutter-6 found: NO (tried pkgconfig and cmake)

meson.build:40:2: ERROR: Dependency "mutter-clutter-6" not found, tried pkgconfig and cmake

A full log can be found at /var/tmp/portage/x11-misc/gpaste-3.36.3/work/gpaste-3.36.3-build/meson-logs/meson-log.txt
 * ERROR: x11-misc/gpaste-3.36.3::testworld failed (configure phase):
 *   (no error message)
```

zigford replied:
Perhaps the old gpaste ebuild was wrong, or perhaps gpaste upstream moved to using mutter over clutter libs at some point, but replacing clutter dep with mutter has resolved this issue for me in my testing. However, I was only able to reproduce this issue when enabling the introspection use flag.
